apps/remote_control/lib/lexical/remote_control/api/proxy/state.ex

```diff
@@ -0,0 +1,146 @@
+defmodule Lexical.RemoteControl.Api.Proxy.State do
+  alias Lexical.Document
+  alias Lexical.RemoteControl
+  alias Lexical.RemoteControl.Api
+  alias Lexical.RemoteControl.Build
+  alias Lexical.RemoteControl.Commands
+
+  import Api.Messages
+  import Api.Proxy.Records
+
+  defstruct initiator_pid: nil, buffer: []
+
+  def new(initiator_pid) do
+    %__MODULE__{initiator_pid: initiator_pid}
+  end
+
+  def add_mfa(%__MODULE__{} = state, mfa() = mfa_record) do
+    %__MODULE__{state | buffer: [mfa_record | state.buffer]}
+  end
+
+  def flush(%__MODULE__{} = state) do
+    {messages, commands} =
+      state.buffer
+      |> Enum.reverse()
+      |> Enum.split_with(fn value ->
+        match?(mfa(module: RemoteControl.Dispatch, function: :broadcast), value)
+      end)
+
+    {project_compile, document_compiles, reindex} = collapse_commands(commands)
+
+    all_commands = [project_compile | Map.values(document_compiles)]
+
+    all_commands
+    |> Enum.concat(collapse_messages(messages, project_compile, document_compiles))
+    |> Enum.filter(&match?(mfa(), &1))
+    |> Enum.sort_by(fn mfa(seq: seq) -> seq end)
+    |> then(fn commands ->
+      if reindex do
+        commands ++ [reindex]
+      else
+        commands
+      end
+    end)
+  end
+
+  defp collapse_commands(commands) do
+    # Rules for collapsing commands
+    # 1. If there's a project compilation requested, remove all document compilations
+    # 2. Formats can be dropped, as they're only valid for a short time.
+    # 3. If there's a reindex, do it after the project compilation has finished
+
+    initial_state = %{project_compiles: [], document_compiles: %{}, reindex: nil}
+
+    grouped =
+      commands
+      |> Enum.reduce(
+        initial_state,
+        fn
+          mfa(module: Build, function: :schedule_compile) = mfa, acc ->
+            Map.update(acc, :project_compiles, [mfa], &[mfa | &1])
+
+          mfa(module: Build, function: :compile_document) = mfa, acc ->
+            mfa(arguments: [_, document]) = mfa
+            uri = document.uri
+            put_in(acc, [:document_compiles, uri], mfa)
+
+          mfa(module: Commands.Reindex) = mfa, acc ->
+            Map.put(acc, :reindex, mfa)
+
+          _, acc ->
+            acc
+        end
+      )
+
+    %{
+      project_compiles: project_compiles,
+      document_compiles: document_compiles,
+      reindex: reindex
+    } = grouped
+
+    project_compile =
+      Enum.reduce(project_compiles, nil, fn
+        mfa(arguments: [_, true]) = mfa, _ ->
+          mfa
+
+        mfa(arguments: [true]) = mfa, _ ->
+          mfa
```

Comment:
Looking at `Build.schedule_compile/2`, the signature is `schedule_compile(project, force? \\ false)`, so I think this case shouldn't ever be hit?

Reply:
so that's actually two MFAs, and I think both need to be accounted for, no?